modules/client-handler/src/main/java/org/apache/ignite/client/handler/JdbcQueryEventHandlerImpl.java

```diff
@@ -178,69 +169,83 @@ private QueryContext createQueryContext(JdbcStatementType stmtType) {
+                        list.add(cnt > Integer.MAX_VALUE ? Statement.SUCCESS_NO_INFO : cnt.intValue());
+
+                        return list;
+                    }));
         }
 
-        return CompletableFuture.completedFuture(new BatchExecuteResult(res.toIntArray()));
+        return tail.handle((ignored, t) -> {
+            if (t != null) {
+                return handleBatchException(t, queries.get(counters.size()), counters.toIntArray());
+            }
+
+            return new BatchExecuteResult(counters.toIntArray());
+        });
     }
 
     /** {@inheritDoc} */
     @Override
     public CompletableFuture<BatchExecuteResult> batchPrepStatementAsync(BatchPreparedStmntRequest req) {
-        IntList res = new IntArrayList(req.getArgs().size());
+        var argList = req.getArgs();
 
-        QueryContext context = createQueryContext(JdbcStatementType.UPDATE_STATEMENT_TYPE);
+        var counters = new IntArrayList(req.getArgs().size());
+        var tail = CompletableFuture.completedFuture(counters);
 
-        try {
-            for (Object[] arg : req.getArgs()) {
-                executeAndCollectUpdateCount(context, req.schemaName(), req.getQuery(), arg, res);
-            }
-        } catch (Exception e) {
-            return handleBatchException(e, req.getQuery(), res);
+        for (Object[] args : argList) {
+            tail = tail.thenCompose(list -> executeAndCollectUpdateCount(req.schemaName(), req.getQuery(), args)
+                    .thenApply(cnt -> {
+                        list.add(cnt > Integer.MAX_VALUE ? Statement.SUCCESS_NO_INFO : cnt.intValue());
+
+                        return list;
+                    }));
         }
 
-        return CompletableFuture.completedFuture(new BatchExecuteResult(res.toIntArray()));
+        return tail.handle((ignored, t) -> {
+            if (t != null) {
+                return handleBatchException(t, req.getQuery(), counters.toIntArray());
+            }
+
+            return new BatchExecuteResult(counters.toIntArray());
+        });
     }
 
-    private void executeAndCollectUpdateCount(QueryContext context, String schema, String sql, Object[] arg, IntList res) {
-        List<SqlCursor<List<?>>> cursors = processor.query(context, schema, sql, arg);
-        for (SqlCursor<List<?>> cursor : cursors) {
-            long updatedRows = (long) cursor.next().get(0);
-            res.add((int) updatedRows);
+    private CompletableFuture<Long> executeAndCollectUpdateCount(String schema, String sql, Object[] arg) {
+        var context = createQueryContext(JdbcStatementType.UPDATE_STATEMENT_TYPE);
+
+        var cursors = processor.queryAsync(context, schema, sql, arg);
+
+        if (cursors.size() != 1) {
+            return CompletableFuture.failedFuture(new IgniteInternalException(""));
         }
+
+        return cursors.get(0).thenCompose(cursor -> cursor.requestNext(1).thenApply(batch -> (Long) batch.items().get(0).get(0)));
```

Comment:
In what thread the request `cursor.requestNext(1)` will be sent?
Is there any possibility that the thread will stuck on sending and block incoming messages?

Reply:
> In what thread the request cursor.requestNext(1) will be sent?

it depends on the particular implementation. For now every such request will be passed to the `QueryTaskExecutor` which processes actual query execution

Reply:
> Is there any possibility that the thread will stuck on sending and block incoming messages?

Only if I made mistake somewhere